File: frontend/js/screens/returns.js
/* returns.js — 18: the return-to-shelf list.
 *
 * Units from cancelled orders. A pick took them off the ledger, so until
 * somebody scans each one back at the rack they are stock nobody can sell.
 * Oldest first; anyone on shift can work it (PRD decision: return by anyone).
 */
(function () {
  'use strict';
  const W = NJW.wire;
  const { region, setF, esc, biAttr, applyLangTo, fail, codeHtml } = W;

  function since(sec) {
    if (sec == null) return '';
    const m = Math.max(0, Math.round(sec / 60));
    if (m < 60) return m + ' mnt';
    return Math.floor(m / 60) + ' jam ' + String(m % 60).padStart(2, '0') + ' mnt';
  }

  function row(t) {
    const left = t.qty - t.qty_returned;
    return '<div class="task task--low" data-task-id="' + t.id + '">' +
      '<img class="task__photo" src="assets/products/placeholder.svg" data-photo-key="' +
        esc((t.brand_sku_code || '').toLowerCase()) + '" alt="">' +
      '<span class="task__ident">' +
        '<span class="task__name">' + esc(t.sku_name) + '</span>' +
        '<span class="task__sku">' + esc(t.brand_sku_code || '') +
          (t.qty_returned ? ' · <span ' + biAttr('sudah kembali', 'already back') +
            '>sudah kembali</span> ' + t.qty_returned + ' / ' + t.qty : '') + '</span>' +
      '</span>' +
      '<span class="task__move">' +
        '<span class="task__loc">' +
          '<span class="task__loc-label" ' + biAttr('Dari pesanan batal', 'From cancelled order') +
            '>Dari pesanan batal</span>' +
          '<span class="task__loc-code">' + esc(t.external_ref || '—') + '</span>' +
        '</span>' +
        '<span class="task__arrow" aria-hidden="true">→</span>' +
        '<span class="task__loc">' +
          '<span class="task__loc-label" ' + biAttr('Ke rak', 'To rack') + '>Ke rak</span>' +
          '<span class="task__loc-code">' + (t.location_code ? codeHtml(t.location_code) :
            '<span ' + biAttr('Belum ada rak', 'No rack yet') + '>Belum ada rak</span>') + '</span>' +
        '</span>' +
      '</span>' +
      '<span class="task__qty">' +
        '<span class="task__qty-num">' + left + '</span>' +
        '<span class="task__qty-unit" ' + biAttr('unit', 'units') + '>unit</span>' +
      '</span>' +
      '<span class="task__since">' +
        '<span class="tasktag tasktag--low" ' + biAttr('Dibatalkan', 'Cancelled') + '>Dibatalkan</span>' +
        '<span class="note" style="font-size:13px">' + since(t.age_seconds) + '</span>' +
      '</span>' +
      '<a class="btn btn--primary" href="18-kembalikan-pindai.html?task=' + t.id + '" ' +
        biAttr('Kerjakan', 'Do it') + '>Kerjakan</a>' +
    '</div>';
  }

  NJW.screens.returns = async () => {
    const site = W.site();
    const list = region('tasks'), empty = region('empty');

    async function load() {
      const r = await NJW.api.returns({ site_id: site.id, status: 'open' });
      setF('task-count', r.open_count);
      setF('unit-count', r.tasks.reduce((n, t) => n + (t.qty - t.qty_returned), 0));
      list.innerHTML = r.tasks.map(row).join('');
      list.hidden = !r.tasks.length;
      if (empty) empty.hidden = !!r.tasks.length;
      applyLangTo(list);
      NJW.paintPhotos(list);
    }

    try { await load(); } catch (e) { fail(e); }
    // Cancellations arrive from Hiryu at any time; the list keeps itself fresh.
    setInterval(() => { if (!document.hidden) load().catch(() => {}); }, 15000);
  };
})();

File: backend/routers/returns.py
"""Return to shelf — units from cancelled orders going back on the rack.

Canonical design: when Hiryu cancels (message 2) the WMS releases what was only
allocated, and anything already picked goes to return-to-shelf. A pick removed
those units from the ledger when they were scanned into the tote, so they come
back the same way: one scan per unit at the rack, each scan a `return_in`
movement. Every movement already triggers message 3, so Hiryu hears the stock
come back the moment it is really on the shelf, not before.

Anyone on shift can work this list. The scan is still verified: a unit that is
not the SKU on the task is refused, exactly like a wrong pick.
"""
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, Query

import auth
import common
import db
import ledger
import models

router = APIRouter(prefix="/api", tags=["returns"])

_SELECT = (
    "SELECT rt.*, s.name_display AS sku_name, s.brand_sku_code, "
    "       l.code AS location_code "
    "FROM return_tasks rt JOIN skus s ON s.id = rt.sku_id "
    "LEFT JOIN locations l ON l.id = rt.location_id "
)


def _out(row: dict) -> dict:
    created = row.get("created_at")
    age = None
    if created is not None:
        # Naive DATETIMEs in this database are UTC (db.connect pins the session).
        age = int((datetime.now(timezone.utc)
                   - created.replace(tzinfo=timezone.utc)).total_seconds())
    return {
        "id": row["id"], "site_id": row["site_id"], "sku_id": row["sku_id"],
        "sku_name": row["sku_name"], "brand_sku_code": row.get("brand_sku_code"),
        "external_ref": row.get("external_ref"),
        "location_id": row.get("location_id"), "location_code": row.get("location_code"),
        "qty": row["qty"], "qty_returned": row["qty_returned"],
        "reason": row["reason"], "status": row["status"],
        "created_at": str(created) if created else None,
        "age_seconds": age,
    }


async def create_for_cancel(cur, *, order: dict, lines: list[dict]) -> int:
    """Queue every picked unit of a cancelled order for return. Returns units queued.

    Called inside the cancel transaction, so the order cannot end up cancelled
    with its picked units forgotten. Units go back where they were picked from;
    when that is unknown, to the SKU's rack.
    """
    queued = 0
    for l in lines:
        picked = int(l.get("qty_picked") or 0)
        if picked <= 0:
            continue
        location_id = l.get("location_id")
        if not location_id:
            slot = await common.slot_for(order["site_id"], l["sku_id"])
            location_id = slot["location_id"] if slot else None
        await db.run(
            cur,
            "INSERT INTO return_tasks (site_id, sku_id, order_id, external_ref, "
            "location_id, qty, reason, is_training) "
            "VALUES (%s,%s,%s,%s,%s,%s,'cancelled',%s)",
            (order["site_id"], l["sku_id"], order["id"], order["external_ref"],
             location_id, picked, 1 if order.get("is_training") else 0),
        )
        queued += picked
    return queued


@router.get("/returns", response_model=models.ReturnTaskList)
async def list_returns(
    site_id: int,
    status: str = Query(default="open", pattern="^(open|done|all)$"),
    limit: int = Query(default=100, ge=1, le=500),
    user: auth.User = Depends(auth.current_user),
):
    await auth.assert_site_access(user, site_id)
    where, params = "WHERE rt.site_id = %s", [site_id]
    if status != "all":
        where += " AND rt.status = %s"
        params.append(status)
    # Oldest first: a unit sitting in a tote is stock nobody can sell.
    rows = await db.fetch_all(
        _SELECT + where + " ORDER BY rt.created_at ASC, rt.id ASC LIMIT %s",
        (*params, limit),
    )
    open_count = await db.fetch_one(
        "SELECT COUNT(*) AS n FROM return_tasks WHERE site_id = %s AND status = 'open'",
        (site_id,),
    )
    return {"tasks": [_out(r) for r in rows], "open_count": int(open_count["n"])}


@router.post("/returns/{task_id}/scan", response_model=models.ReturnScanResult)
async def scan_return(
    task_id: int,
    body: models.ReturnScanIn,
    user: auth.User = Depends(auth.current_user),
):
    """One unit back on the shelf. Scan-verified, no override."""
    replayed = await ledger.replay(body.idempotency_key, "return_scan")
    if replayed:
        return replayed

    task = await db.fetch_one(_SELECT + "WHERE rt.id = %s", (task_id,))
    if not task:
        raise HTTPException(404, "Return task not found")
    site = await auth.assert_site_access(user, task["site_id"])
    if task["status"] != "open":
        raise HTTPException(409, "Sudah selesai dikembalikan. / Already returned.")
    if not task["location_id"]:
        raise HTTPException(
            409, "Barang ini belum punya rak. Panggil supervisor. / "
                 "This SKU has no rack yet. Call a supervisor.")

    code = body.code.strip()
    plate = None
    sku = await common.sku_by_barcode(code)
    if not sku:
        plate = await common.plate_by_code(code)
        if plate and plate["sku_id"]:
            sku = {"id": plate["sku_id"]}
    if not sku:
        raise HTTPException(422, "Barcode tidak dikenal. / Unknown barcode.")
    if sku["id"] != task["sku_id"]:
        raise HTTPException(
            409, f"Salah barang. Yang dikembalikan: {task['sku_name']}. / "
                 f"Wrong item. This return is {task['sku_name']}.")
    if plate and plate["state"] != "picked":
        raise HTTPException(
            409, "Label ini tidak sedang di luar rak. / This label is not out of the rack.")

    async with db.tx() as cur:
        # Re-read under lock: two people scanning the same tote must not
        # return more units than the order took.
        locked = await db.one(
            cur, "SELECT qty, qty_returned, status FROM return_tasks "
                 "WHERE id = %s FOR UPDATE", (task_id,))
        if locked["status"] != "open" or locked["qty_returned"] >= locked["qty"]:
            raise HTTPException(409, "Sudah selesai dikembalikan. / Already returned.")

        await ledger.apply(
            cur, site_id=task["site_id"], sku_id=task["sku_id"],
            location_id=task["location_id"], qty_delta=1,
            movement_type="return_in", actor_email=user.email,
            ref_type="return_task", ref_id=task_id,
            plate_id=plate["id"] if plate else None,
            scan_source="plate" if plate else "scan",
            is_training=bool(site["is_training"]),
        )
        if plate:
            await db.run(
                cur, "UPDATE unit_plates SET state='in_stock', location_id=%s, "
                     "last_seen_at=NOW() WHERE id = %s",
                (task["location_id"], plate["id"]))

        returned = locked["qty_returned"] + 1
        done = returned >= locked["qty"]
        await db.run(
            cur,
            "UPDATE return_tasks SET qty_returned = %s, status = %s, "
            "done_at = " + ("NOW()" if done else "NULL") + ", done_by = %s WHERE id = %s",
            (returned, "done" if done else "open", user.email if done else None, task_id),
        )
        task.update(qty_returned=returned, status="done" if done else "open")
        result = {
            "ok": True, "task": _out(task), "done": done,
            "message": ("Selesai. Semua sudah kembali di rak. / Done. All back on the rack."
                        if done else
                        f"{returned} dari {locked['qty']}. / {returned} of {locked['qty']}."),
        }
        await ledger.remember(cur, body.idempotency_key, "return_scan", result)
    return result
